client/txn.go

```
@@ -356,99 +349,82 @@ func endTxnCall(commit bool, hasTrigger bool) proto.Call {
 	for r := retry.Start(txn.db.txnRetryOptions); r.Next(); {
-		txn.haveTxnWrite, txn.haveEndTxn = false, false // always reset before [re]starting txn
 		if err = retryable(txn); err == nil {
-			if !txn.haveEndTxn {
-				// If there were no errors running retryable, commit the txn. This
-				// may block waiting for outstanding writes to complete in case
-				// retryable didn't -- we need the most recent of all response
-				// timestamps in order to commit.
-				err = txn.Commit()
-			}
+			// If there were no errors running retryable, commit the txn. This
+			// may block waiting for outstanding writes to complete in case
+			// retryable didn't -- we need the most recent of all response
+			// timestamps in order to commit.
+			err = txn.commit()
 		}
 		if restartErr, ok := err.(proto.TransactionRestartError); ok {
 			if log.V(2) {
 				log.Warning(err)
 			}
-			if restartErr.CanRestartTransaction() == proto.TransactionRestart_IMMEDIATE {
+			switch restartErr.CanRestartTransaction() {
+			case proto.TransactionRestart_IMMEDIATE:
 				r.Reset()
 				continue
-			} else if restartErr.CanRestartTransaction() == proto.TransactionRestart_BACKOFF {
+			case proto.TransactionRestart_BACKOFF:
 				continue
 			}
 			// By default, fall through and break.
 		}
 		break
 	}
-	if err != nil {
-		// If the retry logic gave up and haveEndTxn is true, then even if we
-		// tried to run EndTransaction, it must have failed (or was never run;
-		// after all, it's always the last call to be executed). So we pretend
-		// we never sent one (this is necessary to send another one).
-		txn.haveEndTxn = false // if we sent one, it didn't succeed
-		if replyErr := txn.Rollback(); replyErr != nil {
-			log.Errorf("failure aborting transaction: %s; abort caused by: %s", replyErr, err)
-		}
-	}
-	return
+	txn.cleanup(err)
+	return err
 }
 
 // send runs the specified calls synchronously in a single batch and
 // returns any errors. If the transaction is read-only, a potential
```

// If the transaction is read-only or has already been successfully committed, a potential trailing `EndTransaction` call is silently dropped, allowing the caller to always commit or clean-up explicitly even when that may not be required (or even erroneous).